### ISSUE TITLE
security: allow users to enable password auth for 'root'

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#43847]
tldr; this issue proposes to enable:

- setting/changing the password of the `root` user;
- allowing root to log in using their password on the UI;
- allowing root to log in using their password on SQL, subject to regular rules in the HBA configuration.

It also fixes the [regression](https://github.com/cockroachdb/cockroach/issues/43847#issuecomment-572615646)  introduced by #42563, by enabling `root` authentication on the admin UI to obtain a login cookie.

### Background

CockroachDB currently offers 5 separate guardrails to ensure that `root` is always able to connect even when the authentication configuration is botched:

1. when running `--insecure`, anyone can log in without auth.
2. the crdb [HBA configuration](https://www.postgresql.org/docs/current/auth-pg-hba-conf.html) enforces that `host all root all cert` is always the first rule, meaning that clients can log in as `root` if and only if they present a valid TLS client cert for `root` (and because the method is `cert` and not `cert-password`, password auth is rejected for `root` in any case).
3. `ALTER USER WITH PASSWORD` is disallowed for `root`, so root cannot have a password
4. the internal "check password" mechanism, shared by both SQL and HTTP connections, to compare a client-provided password with a user record fails with an error if the user is `root`.
5. the UI `UserLogin` HTTP API reports an error if the user is `root`.
6. the cockroach CLI commands using SQL connections report an error if the user is `root` and a password is supplied.

Meanwhile, there is a separate, unrelated (but relevant) rule:

7. if a user's password is unset/empty, then this user is not able to use password authentication.

### Proposal

This issue proposes to **remove rules 3-6**  specifically, without changing the others.

This proposal would not change the security rules for a cluster using the default configuration: by default, root would not be able to log in using password anywhere (by default, the root account has no password so rule 6 applies) and is required to present a cert on SQL (due to rule 2).

### Non-Pitfalls

A possible counter-argument to this proposal is that CockroachdB uses `root` internally to establish SQL client connections towards itself. 

This is not an obstacle: in insecure clusters, the password would be ignored anyway; in secure clusters, CockroachDB's internal connections use a valid client cert.

**Comment by knz** [comment]
@piyush-singh I am putting this on the docket because addressing this would simplify the product and make it easier to document / reason about it. 

@aaron-crl can you check the proposal?

**Comment by kenliu** [comment]
@knz please make sure that @rolandcrosby is aware of any security related feature proposals as he is responsible for this area.

**Comment by knz** [comment]
oh sorry I got the PMs mixed up. thanks for the reminder.

**Comment by aaron-crl** [comment]
Three quick questions:

(1) Was there a reason we did this as certificate only access?
(2) Would it be possible to separate “root” from internal connections and instead use a reserved exclusive account for those?
(3) Why do we have a “root” account? I know this sounds pedantic but what use case(s) does this account currently support (possibly related to question 1).

**Comment by knz** [comment]
(1) the reason is historical: this predefined rule existed before we started to support password auth

(2) we already use another special account `node` for most internal connections. The only few that use SQL could be migrated to use a separate special account I guess.

(3) the `root` account is the one account with the `admin` role available in a newly created cluster. The admin role is necessary to create other users, grant privileges etc.

Pursuant to (3) I guess one could ask "what makes `root` special and why could it not become a regular user account" -- the reason is that `root` cannot be deleted from the cluster; we have this as ground rule to prevent users from shooting themselves in the foot and becoming unable to administrate their cluster.

**Comment by aaron-crl** [comment]
Thanks!

Sounds good to me with the caveat that it would be nice to see all cluster internal functions use their own security principals to make security logging and eventing easier in the future re: (2).

**Comment by knz** [comment]
> all cluster internal functions use their own security principals

That's quite an interesting idea.

FYI for all internal uses of SQL we have this already, because SQL connections have a special token called `application_name` which is used in logging and our internal uses already populate this with a different value each time.

For RPC connections I guess we could consider this as an addition. Seems orthogonal though.

**Comment by mattcrdb** [comment]
Our recent security fix https://github.com/cockroachdb/cockroach/pull/42563  is preventing access to the HTTP endpoints in non-enterprise clusters. Specifically the `#/node/x/logs` endpoint requires root user access or the admin-role enabled on the sql user. Without an enterprise licence you can only view that endpoint as the root user (as granting roles is an enterprise feature), however because the root user can't have a password - that page and any other pages that quire this role are blocked.

**Comment by aaron-crl** [comment]
> Seems orthogonal though.
Agreed, we should move that discussion elsewhere. It's a nice to have from a security hygiene standpoint; no other concerns.

**Comment by israellot** [comment]
Allowing to assign the admin role to one additional user would also fix the problem described by @mattcrdb

**Comment by knz** [comment]
Root passwords are only one of the solutions for the regression. I have generalized the statement of the regression in the following issue: #43870
